src/coreclr/jit/llvm.cpp

```diff
@@ -1896,7 +1802,252 @@ void Llvm::ConvertShadowStackLocalNode(GenTreeLclVarCommon* node)
+{
+    GenTree* storeNode;
+    if (nodeType == TYP_STRUCT)
+    {
+        storeNode = new (_compiler, GT_STORE_OBJ)
+            GenTreeObj(nodeType, addr, nodeToStore, structClassLayout);
+    }
+    else
+    {
+        storeNode = new (_compiler, GT_STOREIND) GenTreeStoreInd(nodeType, addr, nodeToStore);
+    }
+    return storeNode;
+}
+
+//------------------------------------------------------------------------
+// lowerCallToShadowStack: Lower the call, rewriting its arguments.
+//
+// This method has two primary objectives:
+//  1) Transfer the information about the arguments from arg info to explicit
+//     PutArgType nodes, to make it easy for codegen to consume it. Also, get
+//     rid of the late/non-late argument distinction, by sorting the inserted nodes
+//     in the original evaluation order, matching that of them in the signature.
+//  2) Rewrite arguments and the return to be stored on the shadow stack. We take
+//     the arguments which need to be on the shadow stack, remove them from the call
+//     arguments list, store their values on the shadow stack, at offsets calculated
+//     in a simple increasing order, matching the signature. We also rewrite returns
+//     that must be on the shadow stack, see "lowerCallReturn".
+//
+void Llvm::lowerCallToShadowStack(GenTreeCall* callNode, CORINFO_SIG_INFO& calleeSigInfo)
+{
+    // rewrite the args, adding shadow stack, and moving gc tracked args to the shadow stack
+    unsigned shadowStackUseOffest = 0;
+
+    fgArgInfo*                 argInfo    = callNode->fgArgInfo;
+    unsigned int               argCount   = argInfo->ArgCount();
+    fgArgTabEntry**            argTable   = argInfo->ArgTable();
+    std::vector<OperandArgNum> sortedArgs = std::vector<OperandArgNum>(argCount);
+    OperandArgNum*             sortedData = sortedArgs.data();
+
+    GenTreeCall::Use* lastArg;
+    GenTreeCall::Use* insertReturnAfter;
+    GenTreeCall::Use* callThisArg = callNode->gtCallThisArg;
+
+    callNode->ResetArgInfo();
+    callNode->gtCallThisArg = nullptr;
+    argCount                = argInfo->ArgCount();
+    argTable                = argInfo->ArgTable();
+
+    // set up the callee shadowstack, creating a temp and the PUTARG
+    GenTreeLclVar* shadowStackVar = _compiler->gtNewLclvNode(_shadowStackLclNum, TYP_I_IMPL);
+    GenTreeIntCon* offset         = _compiler->gtNewIconNode(_shadowStackLocalsSize,
+                                                     TYP_I_IMPL); // TODO-LLVM: possible performance benefit: when
+                                                                  // _shadowStackLocalsSize == 0, then omit the GT_ADD
+    GenTree* calleeShadowStack = _compiler->gtNewOperNode(GT_ADD, TYP_I_IMPL, shadowStackVar, offset);
+
+    GenTreePutArgType* putArg = _compiler->gtNewPutArgType(TYP_I_IMPL, calleeShadowStack, CORINFO_TYPE_PTR, NO_CLASS_HANDLE);
```

Comment:
Nit: let's rename `putArg` as something like `calleeShadowStackPutArg`, to not have redeclared locals (there is another `GenTreePutArgType* putArg` below).